tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+                    ],
+                ],
+            ],
+            'lastPrice' => ['$last' => '$prices'],
+        ],
+    ],
+];
+```
+
+Once again, the builder can make this a little more concise, but the complexity
+remains:
+
+```php
+$prices = Expression::arrayFieldPath('prices');
+$reportDate = Expression::variable('priceReport.reportDate');
+$previousReportDate = Expression::variable('priceReport.previous.reportDate');
+
+$pipeline = new Pipeline(
+    Stage::addFields(
+        weightedPrices: Expression::map(
+            input: $prices,
+            in: object(
+                duration: Expression::min(
+                    Expression::dateDiff(
+                        startDate: $previousReportDate,
+                        endDate: $reportDate,
+                        unit: 'second',
+                    ),
+                    Expression::add(
+                        Expression::multiply(
+                            Expression::hour($reportDate),
+                            3600,
+                        ),
+                        Expression::multiply(
+                            Expression::minute($reportDate),
+                            60,
+                        ),
+                        Expression::second($reportDate),
+                    )
+                ),
+                price: Expression::variable('priceReport.previous.price'),
+            ),
+            as: 'priceReport',
+        ),
+        lastPrice: Expression::last($prices),
+    ),
+);
+```
+
+The main cognitive effort comes from computing the `duration` field. So, this
+should be our focus. Together with some comments, we can extract the logic to
+separate builder methods:
+
+```php
+public static function computeElapsedSecondsOnDay(
+    Expression\ResolvesToDate $date,
+): Expression\ResolvesToInt {
+    return Expression::add(
+        Expression::multiply(
+            Expression::hour($date),
+            3600,
+        ),
+        Expression::multiply(
+            Expression::minute($date),
+            60,
+        ),
+        Expression::second($date),
+    );
+}
+
+/**
+ * Returns the time that has elapsed between two dates, in seconds.
+ *
+ * If the dates are on the same day, the difference is computed between the
+ * two times. If the dates fall on different dates, the duration returned is
+ * the number of seconds since the start of the day from the second date
+ * argument.
+ */
+public static function computeDurationBetweenDates(
+    Expression\ResolvesToDate $previousReportDate,
+    Expression\ResolvesToDate $reportDate,
+): Expression\ResolvesToInt {
+    return Expression::min(
+        Expression::dateDiff(
+            startDate: $previousReportDate,
+            endDate: $reportDate,
+            unit: 'second',
+        ),
+        self::computeElapsedSecondsOnDay($reportDate),
+    );
+}
+```
+
+Again, this reduces the complexity of the pipeline stage tremendously:
```

Comment:
```suggestion
This reduces the complexity of the pipeline stage tremendously:
```

Removing more filler words. Fine in a conference talk, but I don't think they're needed in a blog post.